File: Sources/UIViewController/UIViewController+XIBCreation.m
//
//  UIViewController+XIBCreation.m
//  SmartCity
//
//  Created by ZCGC on 2017/5/12.
//  Copyright © 2017年 NRH. All rights reserved.
//

#import "UIViewController+XIBCreation.h"
#import <objc/runtime.h>

@implementation UIViewController (XIBCreation)

+ (instancetype)loadFromNib {
    NSString *nibName = NSStringFromClass([self class]);
    NSString *path = [[NSBundle mainBundle] pathForResource:nibName ofType:@"nib"];
    if (!path) {
        return [[self alloc] init];
    }
    return [[self alloc] initWithNibName:nibName bundle:nil];
}

@end


@interface UIViewController (PresentationModal)
@end
@implementation UIViewController (PresentationModal)

+ (void)load {
    SEL originalSelector = @selector(init);
    SEL swizzledSelector = @selector(hook_init);
    
    Method originalMethod = class_getInstanceMethod(self, originalSelector);
    Method swizzledMethod = class_getInstanceMethod(self, swizzledSelector);
    
    BOOL didAddMethod =
    class_addMethod(self,
                    originalSelector,
                    method_getImplementation(swizzledMethod),
                    method_getTypeEncoding(swizzledMethod));
    
    if (didAddMethod) {
        class_replaceMethod(self,
                            swizzledSelector,
                            method_getImplementation(originalMethod),
                            method_getTypeEncoding(originalMethod));

    } else {
        method_exchangeImplementations(originalMethod, swizzledMethod);
    }
}

- (instancetype)hook_init {
    id instance = [self hook_init];
    self.modalPresentationStyle = UIModalPresentationFullScreen;
    self.modalPresentationCapturesStatusBarAppearance = YES;
    return instance;
}

@end

@interface UINavigationController (PresentationModal)
@end

@implementation UINavigationController (PresentationModal)

+ (void)load {
    SEL originalSelector = @selector(initWithRootViewController:);
    SEL swizzledSelector = @selector(hook_initWithRootViewController:);
    
    Method originalMethod = class_getInstanceMethod(self, originalSelector);
    Method swizzledMethod = class_getInstanceMethod(self, swizzledSelector);
    
    BOOL didAddMethod =
    class_addMethod(self,
                    originalSelector,
                    method_getImplementation(swizzledMethod),
                    method_getTypeEncoding(swizzledMethod));
    
    if (didAddMethod) {
        class_replaceMethod(self,
                            swizzledSelector,
                            method_getImplementation(originalMethod),
                            method_getTypeEncoding(originalMethod));
        
    } else {
        method_exchangeImplementations(originalMethod, swizzledMethod);
    }
}

- (instancetype)hook_initWithRootViewController:(UIViewController *)rootViewController {
    id instance = [self hook_initWithRootViewController:rootViewController];
    self.modalPresentationStyle = UIModalPresentationFullScreen;
    self.modalPresentationCapturesStatusBarAppearance = YES;
    return instance;
}

@end

@interface UIStoryboard (PresentationModal)
@end
@implementation UIStoryboard (PresentationModal)
+ (void)load {
    SEL originalSelector = @selector(instantiateInitialViewController);
    SEL swizzledSelector = @selector(hook_instantiateInitialViewController);
    
    Method originalMethod = class_getInstanceMethod(self, originalSelector);
    Method swizzledMethod = class_getInstanceMethod(self, swizzledSelector);
    
    BOOL didAddMethod =
    class_addMethod(self,
                    originalSelector,
                    method_getImplementation(swizzledMethod),
                    method_getTypeEncoding(swizzledMethod));
    
    if (didAddMethod) {
        class_replaceMethod(self,
                            swizzledSelector,
                            method_getImplementation(originalMethod),
                            method_getTypeEncoding(originalMethod));
        
    } else {
        method_exchangeImplementations(originalMethod, swizzledMethod);
    }
}

- (UIViewController *)hook_instantiateInitialViewController {
    UIViewController * vc = [self hook_instantiateInitialViewController];
    vc.modalPresentationStyle = UIModalPresentationFullScreen;
    vc.modalPresentationCapturesStatusBarAppearance = YES;
    return vc;
}
@end

File: Sources/UIViewController/UIViewController+XIBCreation.swift
//
//  UIViewController+XIBCreation.swift
//  SPUIKitCategory module
//
//  Created by LSP on 2020/12/22.
//  Copyright © 2020 LSP. All rights reserved.
//

#if !os(macOS)
import UIKit

@objc
extension UIViewController {
    /// 需要交换的方法名
    var UIViewControllerXibCreationExchangeMethodNames: [String] {
        return ["init"]
    }
    
    func sp_init() -> UIViewController {
        let vc = self.sp_init()
        vc.modalPresentationStyle = .fullScreen;
        vc.modalPresentationCapturesStatusBarAppearance = true
        return vc;
    }
}

extension UIViewController {
    
    public enum NibLoadType {
        case nib(_ name: String = "")
        case storyboard(_ name: String = "", _ identifier: String = "")
    }
    
    /**
     *  从XIB文件（与类名相同的xib）初始化。
     */
    public
    class func load(from type: NibLoadType, bundle: Bundle = .main) -> Self {
        switch type {
        case .nib(let nib):
            let name = nib.isEmpty ? String(describing: self) : nib
            if let _ = Bundle.main.path(forResource: name, ofType: "xib") {
                return self.init(nibName: name, bundle: bundle)
            }
            break
        case .storyboard(var name, var identy):
            if name.isEmpty {
                name = "Main"
            }
            if identy.isEmpty {
                identy = self.storyboardIndentifier
            }
            if identy.isEmpty {
                identy = String(describing: self)
            }
            if let vc = UIStoryboard(name: name, bundle: bundle).instantiateViewController(withIdentifier: identy) as? Self {
                return vc
            }
        }
        return self.init()
    }
    
    open class var storyboardName: String {
        return "Main"
    }
    
    open class var storyboardIndentifier: String {
        return String(describing: self)
    }
}

@objc
extension UINavigationController {
    /// 需要交换的方法名
    var UINavigationControllerXibCreationExchangeMethodNames: [String] {
        return ["initWithRootViewController:"]
    }
    
    func sp_initWithRootViewController(_ vc: UIViewController) -> UINavigationController {
        let navi = self.sp_initWithRootViewController(vc)
        navi.modalPresentationStyle = .fullScreen
        navi.modalPresentationCapturesStatusBarAppearance = true
        return navi
    }
}

@objc
extension UIStoryboard {
    
    /// 需要交换的方法名
    var UIStoryboardXibCreationExchangeMethodNames: [String] {
        return ["instantiateInitialViewController"]
    }
    
    func sp_instantiateInitialViewController() -> UIViewController? {
        let vc = self.sp_instantiateInitialViewController()
        vc?.modalPresentationStyle = .fullScreen
        vc?.modalPresentationCapturesStatusBarAppearance = true
        return vc
    }
    
    public static var main: UIStoryboard {
        return UIStoryboard(name: "Main", bundle: nil)
    }
}
#endif
